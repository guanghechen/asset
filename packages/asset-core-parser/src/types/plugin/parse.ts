import type { IAssetId } from '@guanghechen/asset-core'

export interface IAssetParserPluginParseApi {
  /**
   * Load source content.
   * @param relativeSrcLocation
   */
  loadContent(relativeSrcLocation: string): Promise<Buffer | null>
  /**
   * Load source content (synchronously).
   * @param relativeSrcLocation
   */
  loadContentSync(relativeSrcLocation: string): Buffer | null
  /**
   * Resolve asset slug.
   * @param slug
   */
  resolveSlug(slug: string | null | undefined): string | null
}

export interface IAssetParserPluginParseNext {
  (embryo: Readonly<IAssetParserPluginParseOutput> | null):
    | IAssetParserPluginParseOutput
    | null
    | Promise<IAssetParserPluginParseOutput | null>
}

export interface IAssetParserPluginParse {
  (
    input: Readonly<IAssetParserPluginParseInput>,
    embryo: Readonly<IAssetParserPluginParseOutput> | null,
    api: Readonly<IAssetParserPluginParseApi>,
    next: IAssetParserPluginParseNext,
  ): IAssetParserPluginParseOutput | null | Promise<IAssetParserPluginParseOutput | null>
}

export interface IAssetParserPluginParseInput {
  /**
   * Asset global unique identifier.
   */
  guid: IAssetId
  /**
   * The fingerprint of the asset content.
   */
  hash: string
  /**
   * Source virtual filepath (*nix style).
   */
  src: string
  /**
   * The created date of the asset (ISOString).
   */
  createdAt: string
  /**
   * The last modification date of the asset (ISOString).
   */
  updatedAt: string
  /**
   * Name of the source file.
   */
  filename: string
  /**
   * Asset title.
   */
  title: string
  /**
   * Raw content.
   */
  content: Buffer
}

export interface IAssetParserPluginParseOutput<D = unknown> {
  /**
   * Asset content type.
   */
  type: string
  /**
   * Asset MIME type.
   */
  mimetype: string
  /**
   * Title of asset.
   */
  title: string
  /**
   * A stable page url to reveal this asset.
   */
  slug: string | null
  /**
   * The created date of the asset (ISOString).
   */
  createdAt: string
  /**
   * The last modification date of the asset (ISOString).
   */
  updatedAt: string
  /**
   * Asset categories, each element represent a category path.
   */
  categories: string[][]
  /**
   * Asset tags.
   */
  tags: string[]
  /**
   * Asset data
   */
  data: D | null
}
