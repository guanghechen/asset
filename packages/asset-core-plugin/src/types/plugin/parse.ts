import type { IAssetId } from '@guanghechen/asset-core'

export interface IAssetPluginParseApi {
  /**
   * Load source content.
   * @param relativeSrcLocation
   */
  loadContent(relativeSrcLocation: string): Promise<Buffer | null>
  /**
   * Load source content (synchronously).
   * @param relativeSrcLocation
   */
  loadContentSync(relativeSrcLocation: string): Buffer | null
  /**
   * Resolve asset slug.
   * @param slug
   */
  resolveSlug(slug: string | null | undefined): string | null
}

export interface IAssetPluginParseNext {
  (embryo: Readonly<IAssetPluginParseOutput> | null):
    | IAssetPluginParseOutput
    | null
    | Promise<IAssetPluginParseOutput | null>
}

export interface IAssetPluginParse {
  (
    input: Readonly<IAssetPluginParseInput>,
    embryo: Readonly<IAssetPluginParseOutput> | null,
    api: Readonly<IAssetPluginParseApi>,
    next: IAssetPluginParseNext,
  ): IAssetPluginParseOutput | null | Promise<IAssetPluginParseOutput | null>
}

export interface IAssetPluginParseInput {
  /**
   * Asset global unique identifier.
   */
  guid: IAssetId
  /**
   * The fingerprint of the asset content.
   */
  hash: string
  /**
   * Source virtual filepath (*nix style).
   */
  src: string
  /**
   * The created date of the asset (ISOString).
   */
  createdAt: string
  /**
   * The last modification date of the asset (ISOString).
   */
  updatedAt: string
  /**
   * Name of the source file.
   */
  filename: string
  /**
   * Asset title.
   */
  title: string
  /**
   * Raw content.
   */
  content: Buffer
}

export interface IAssetPluginParseOutput<D = unknown> {
  /**
   * Asset content type.
   */
  type: string
  /**
   * Asset MIME type.
   */
  mimetype: string
  /**
   * Title of asset.
   */
  title: string
  /**
   * Description of the content.
   */
  description: string | null
  /**
   * A stable page url to reveal this asset.
   */
  slug: string | null
  /**
   * The created date of the asset (ISOString).
   */
  createdAt: string
  /**
   * The last modification date of the asset (ISOString).
   */
  updatedAt: string
  /**
   * Asset categories, each element represent a category path.
   */
  categories: string[][]
  /**
   * Asset tags.
   */
  tags: string[]
  /**
   * Asset data
   */
  data: D | null
}
