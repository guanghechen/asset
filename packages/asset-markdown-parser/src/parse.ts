import type { MdastRoot } from './types/mdast'
import { createExGFMParser } from '@yozora/parser-gfm'

const parser = createExGFMParser({ shouldReservePosition: false })

/**
 * Parse markdown to ast
 *
 * @param content
 */
export const parseToMdast = (content: string): MdastRoot => {
  // resolve content
  const result: MdastRoot = parser.parse(content) as MdastRoot
  return result
}
