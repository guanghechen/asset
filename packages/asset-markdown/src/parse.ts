import gfm from 'remark-gfm'
import math from 'remark-math'
import markdown from 'remark-parse'
import unified from 'unified'
import removePosition from 'unist-util-remove-position'
import { AssetMarkdownEntityContent } from './entity'


export type MarkdownParser<T extends unknown = unknown> =
  (content: string) => AssetMarkdownEntityContent<T>


/**
 * Parse markdown to ast
 *
 * @param rawContent
 */
const processor = unified().use(markdown).use(gfm).use(math)
export const parseMarkdownToAst: MarkdownParser<any> = (rawContent) => {
  // resolve content
  const content: any = processor.runSync(processor.parse(rawContent))
  const summary: any = {
    type: 'root',
    children: content.children.slice(0, 3),
  }

  const result = {
    content: removePosition(content, false),
    summary: removePosition(summary, false),
  }
  return result
}


/**
 * Preserve the raw content
 *
 * @param rawContent
 */
export const preserveMarkdown: MarkdownParser<string> = (rawContent) => {
  return {
    content: rawContent,
    summary: '',
  }
}
