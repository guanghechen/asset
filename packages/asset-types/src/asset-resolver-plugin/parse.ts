import type { IAsset, IAssetMeta } from '../asset'
import type { IBinaryFileData } from '../asset-file'

export interface IAssetPluginParseApi {
  /**
   * Load source content.
   * @param absoluteSrcPath
   */
  loadContent(absoluteSrcPath: string): Promise<IBinaryFileData | null>
  /**
   * Resolve asset by absoluteSrcPath.
   * @param absoluteSrcPath
   */
  resolveAsset(absoluteSrcPath: string): Promise<Readonly<IAsset> | null>
  /**
   * Resolve src path.
   * @param srcPathRelativeToCurDir the path relative to the parent path of the current resource.
   */
  resolveRefPath(srcPathRelativeToCurDir: string): Promise<string | null>
  /**
   * Resolve asset slug.
   * @param meta
   */
  resolveSlug(meta: Readonly<IAssetMeta>): Promise<string | null>
}

export interface IAssetPluginParseNext {
  (embryo: Readonly<IAssetPluginParseOutput> | null): Promise<IAssetPluginParseOutput | null>
}

export interface IAssetPluginParseInput {
  /**
   * Asset source content type.
   */
  sourcetype: string
  /**
   * Asset tittle.
   */
  title: string
  /**
   * Relative path to the source root.
   */
  src: string
  /**
   * File extension (without dot).
   */
  extname: string | undefined
  /**
   * Source file content.
   */
  content: IBinaryFileData
  /**
   * Source file encoding.
   */
  encoding: BufferEncoding | undefined
}

export interface IAssetPluginParseOutput<D = unknown> {
  /**
   * Asset data
   */
  data: D | null
}
