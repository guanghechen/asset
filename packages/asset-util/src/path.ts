import path from 'node:path'

const urlRegex = /^\w+:\/\//

/**
 * Check if the given filepath is an absolute path.
 * @param filepath
 */
export function isAbsolutePath(filepath: string): boolean {
  if (path.isAbsolute(filepath)) return true
  if (urlRegex.test(filepath)) return true
  return false
}
