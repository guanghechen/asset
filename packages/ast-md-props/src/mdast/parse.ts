import gfm from 'remark-gfm'
import math from 'remark-math'
import markdown from 'remark-parse'
import unified from 'unified'
import type { MdastRoot } from './types'


const processor = unified().use(markdown).use(gfm).use(math)


/**
 * Parse markdown to ast
 *
 * @param rawContent
 */
export const parseMdast = (rawContent: string): MdastRoot => {
  // resolve content
  const result: MdastRoot = processor
    .runSync(processor.parse(rawContent)) as MdastRoot
  return result
}
