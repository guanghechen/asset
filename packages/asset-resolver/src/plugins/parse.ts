import type {
  IAsset,
  IAssetParsePlugin,
  IAssetPluginParseApi,
  IAssetPluginParseInput,
  IAssetPluginParseNext,
  IAssetPluginParseOutput,
  IAssetResolverApi,
  IBinaryFileData,
} from '@guanghechen/asset-types'
import path from 'node:path'
import type { IAssetPluginResolveResult } from './resolve'

export async function parse(
  args: IAssetPluginParseArgs,
  plugins: ReadonlyArray<IAssetParsePlugin>,
  api: IAssetResolverApi,
): Promise<IAssetPluginParseResult | null> {
  const { lastStageResult, loadContent, resolveAsset } = args
  const { absoluteSrcPath, src, asset, content, encoding } = lastStageResult
  const input: IAssetPluginParseInput = {
    sourcetype: asset.sourcetype,
    title: asset.title,
    src,
    extname: asset.extname,
    content,
    encoding,
  }

  const curDir: string = path.dirname(absoluteSrcPath)
  const pluginApi: IAssetPluginParseApi = {
    loadContent: async relativePath => {
      const refPath: string | null = api.resolveRefPath(curDir, relativePath)
      if (refPath === null) return null
      return loadContent(refPath)
    },
    resolveAsset: async relativePath => {
      const refPath: string | null = api.resolveRefPath(curDir, relativePath)
      if (refPath === null) return null
      return resolveAsset(refPath)
    },
    resolveSlug: meta => api.uriResolver.resolveSlug(meta),
  }
  const reducer: IAssetPluginParseNext = plugins.reduceRight<IAssetPluginParseNext>(
    (next, middleware) => embryo => middleware.parse(input, embryo, pluginApi, next),
    async embryo => embryo,
  )

  const output: IAssetPluginParseOutput | null = await reducer(null)

  // Don't return null even the output is null cause the plugin could have following stage plugins.
  const result: IAssetPluginParseResult = {
    absoluteSrcPath,
    asset,
    content,
    encoding,
    data: output?.data ?? null,
  }
  return result
}

export interface IAssetPluginParseArgs {
  lastStageResult: IAssetPluginResolveResult
  /**
   * Load content by source file srcPath.
   * @param absoluteSrcPath
   */
  loadContent: (absoluteSrcPath: string) => Promise<IBinaryFileData | null>
  /**
   * Resolve asset by absoluteSrcPath.
   * @param absoluteSrcPath
   */
  resolveAsset(absoluteSrcPath: string): Promise<Readonly<IAsset> | null>
}

export interface IAssetPluginParseResult {
  absoluteSrcPath: string
  asset: IAsset
  content: IBinaryFileData
  encoding: BufferEncoding | undefined
  data: unknown | null
}
