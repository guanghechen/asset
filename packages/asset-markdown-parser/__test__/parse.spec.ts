import fs from 'fs-extra'
import path from 'path'
import parse from '../src'

describe('parser', function () {
  const caseDir = path.resolve(__dirname, 'case')
  const cases = fs.readdirSync(caseDir)
  for (const kase of cases) {
    const filepath = path.join(caseDir, kase)

    test(path.parse(kase).name, async function () {
      const content = await fs.readFile(filepath, 'utf-8')
      const result = parse(content)
      expect(result).toMatchSnapshot()
    })
  }
})
