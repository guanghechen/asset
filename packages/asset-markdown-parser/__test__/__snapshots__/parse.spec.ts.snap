// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parser basic 1`] = `
Object {
  "ast": Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "type": "text",
            "value": "Paragraph",
          },
        ],
        "identifier": "heading-paragraph",
        "level": 1,
        "type": "heading",
      },
      Object {
        "children": Array [
          Object {
            "type": "text",
            "value": "xx
xxx\\\\
xxxxx",
          },
          Object {
            "type": "break",
          },
          Object {
            "type": "text",
            "value": "
xx $x$",
          },
        ],
        "type": "paragraph",
      },
      Object {
        "children": Array [
          Object {
            "type": "text",
            "value": "Emphasis",
          },
        ],
        "identifier": "heading-emphasis",
        "level": 2,
        "type": "heading",
      },
      Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "type": "text",
                "value": "This is an emphasis (italic) paragraph",
              },
            ],
            "type": "emphasis",
          },
        ],
        "type": "paragraph",
      },
      Object {
        "children": Array [
          Object {
            "type": "text",
            "value": "Strong",
          },
        ],
        "identifier": "heading-strong",
        "level": 2,
        "type": "heading",
      },
      Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "type": "text",
                "value": "This is a strong paragraph",
              },
            ],
            "type": "strong",
          },
        ],
        "type": "paragraph",
      },
      Object {
        "children": Array [
          Object {
            "type": "text",
            "value": "Delete",
          },
        ],
        "identifier": "heading-delete",
        "level": 2,
        "type": "heading",
      },
      Object {
        "children": Array [
          Object {
            "type": "text",
            "value": "Well, ",
          },
          Object {
            "children": Array [
              Object {
                "type": "text",
                "value": "this is something going wrong,",
              },
            ],
            "type": "delete",
          },
          Object {
            "type": "text",
            "value": " fine.",
          },
        ],
        "type": "paragraph",
      },
      Object {
        "children": Array [
          Object {
            "type": "text",
            "value": "Code",
          },
        ],
        "identifier": "heading-code",
        "level": 2,
        "type": "heading",
      },
      Object {
        "args": Object {},
        "lang": "javascript",
        "meta": "",
        "type": "code",
        "value": "let x = 1
console.log('x:', x)
",
      },
      Object {
        "args": Object {},
        "lang": "javascript",
        "meta": "literal",
        "type": "code",
        "value": "let x = 1
console.log('x:', x)
",
      },
      Object {
        "args": Object {},
        "lang": "jsx",
        "meta": "live",
        "type": "codeLive",
        "value": "render(
  <div>Hello, world!</div>
)
",
      },
      Object {
        "children": Array [
          Object {
            "type": "text",
            "value": "List",
          },
        ],
        "identifier": "heading-list",
        "level": 2,
        "type": "heading",
      },
      Object {
        "children": Array [
          Object {
            "checked": undefined,
            "children": Array [
              Object {
                "children": Array [
                  Object {
                    "type": "text",
                    "value": "waw",
                  },
                ],
                "type": "paragraph",
              },
              Object {
                "children": Array [
                  Object {
                    "checked": undefined,
                    "children": Array [
                      Object {
                        "type": "text",
                        "value": "foo",
                      },
                    ],
                    "spread": false,
                    "type": "listItem",
                  },
                  Object {
                    "checked": undefined,
                    "children": Array [
                      Object {
                        "type": "text",
                        "value": "bar",
                      },
                    ],
                    "spread": false,
                    "type": "listItem",
                  },
                ],
                "ordered": false,
                "spread": false,
                "start": undefined,
                "type": "list",
              },
            ],
            "spread": false,
            "type": "listItem",
          },
          Object {
            "checked": undefined,
            "children": Array [
              Object {
                "children": Array [
                  Object {
                    "type": "text",
                    "value": "okay",
                  },
                ],
                "type": "paragraph",
              },
            ],
            "spread": false,
            "type": "listItem",
          },
          Object {
            "checked": undefined,
            "children": Array [
              Object {
                "children": Array [
                  Object {
                    "type": "text",
                    "value": "nothing",
                  },
                ],
                "type": "paragraph",
              },
              Object {
                "children": Array [
                  Object {
                    "checked": undefined,
                    "children": Array [
                      Object {
                        "children": Array [
                          Object {
                            "type": "text",
                            "value": "well",
                          },
                        ],
                        "type": "paragraph",
                      },
                    ],
                    "spread": false,
                    "type": "listItem",
                  },
                  Object {
                    "checked": undefined,
                    "children": Array [
                      Object {
                        "children": Array [
                          Object {
                            "type": "text",
                            "value": "banana",
                          },
                        ],
                        "type": "paragraph",
                      },
                    ],
                    "spread": false,
                    "type": "listItem",
                  },
                ],
                "ordered": false,
                "spread": true,
                "start": undefined,
                "type": "list",
              },
            ],
            "spread": false,
            "type": "listItem",
          },
        ],
        "ordered": false,
        "spread": true,
        "start": undefined,
        "type": "list",
      },
      Object {
        "children": Array [
          Object {
            "type": "text",
            "value": "xxx",
          },
        ],
        "identifier": "heading-xxx",
        "level": 4,
        "type": "heading",
      },
      Object {
        "children": Array [
          Object {
            "checked": undefined,
            "children": Array [
              Object {
                "type": "text",
                "value": "waw",
              },
            ],
            "spread": false,
            "type": "listItem",
          },
          Object {
            "checked": undefined,
            "children": Array [
              Object {
                "type": "text",
                "value": "okay",
              },
            ],
            "spread": false,
            "type": "listItem",
          },
          Object {
            "checked": undefined,
            "children": Array [
              Object {
                "type": "text",
                "value": "nothing",
              },
            ],
            "spread": false,
            "type": "listItem",
          },
        ],
        "ordered": true,
        "spread": false,
        "start": 1,
        "type": "list",
      },
      Object {
        "children": Array [
          Object {
            "type": "inlineCode",
            "value": "complex",
          },
          Object {
            "type": "text",
            "value": " ",
          },
          Object {
            "children": Array [
              Object {
                "type": "text",
                "value": "header",
              },
            ],
            "type": "strong",
          },
          Object {
            "type": "text",
            "value": " ~demo~",
          },
        ],
        "identifier": "heading-complex-header-~demo~",
        "level": 3,
        "type": "heading",
      },
      Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "type": "text",
                "value": "Blockquote test",
              },
            ],
            "type": "paragraph",
          },
          Object {
            "children": Array [
              Object {
                "checked": undefined,
                "children": Array [
                  Object {
                    "type": "text",
                    "value": "waw",
                  },
                ],
                "spread": false,
                "type": "listItem",
              },
              Object {
                "checked": undefined,
                "children": Array [
                  Object {
                    "type": "text",
                    "value": "waw",
                  },
                ],
                "spread": false,
                "type": "listItem",
              },
              Object {
                "checked": undefined,
                "children": Array [
                  Object {
                    "type": "text",
                    "value": "waw",
                  },
                ],
                "spread": false,
                "type": "listItem",
              },
              Object {
                "checked": undefined,
                "children": Array [
                  Object {
                    "type": "text",
                    "value": "waw",
                  },
                  Object {
                    "args": Object {},
                    "lang": "javascript",
                    "meta": "",
                    "type": "code",
                    "value": "let x = 1
console.log('x:', x)
",
                  },
                ],
                "spread": false,
                "type": "listItem",
              },
            ],
            "ordered": false,
            "spread": false,
            "start": undefined,
            "type": "list",
          },
        ],
        "type": "blockquote",
      },
      Object {
        "children": Array [
          Object {
            "type": "text",
            "value": "Table",
          },
        ],
        "identifier": "heading-table",
        "level": 1,
        "type": "heading",
      },
      Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "align": undefined,
                "children": Array [
                  Object {
                    "type": "text",
                    "value": "x",
                  },
                ],
                "isHeader": undefined,
                "type": "tableCell",
              },
              Object {
                "align": undefined,
                "children": Array [
                  Object {
                    "type": "text",
                    "value": "y",
                  },
                ],
                "isHeader": undefined,
                "type": "tableCell",
              },
              Object {
                "align": undefined,
                "children": Array [
                  Object {
                    "type": "text",
                    "value": "z",
                  },
                ],
                "isHeader": undefined,
                "type": "tableCell",
              },
            ],
            "type": "tableRow",
          },
          Object {
            "children": Array [
              Object {
                "align": undefined,
                "children": Array [
                  Object {
                    "type": "text",
                    "value": "1",
                  },
                ],
                "isHeader": undefined,
                "type": "tableCell",
              },
              Object {
                "align": undefined,
                "children": Array [
                  Object {
                    "type": "text",
                    "value": "2",
                  },
                ],
                "isHeader": undefined,
                "type": "tableCell",
              },
              Object {
                "align": undefined,
                "children": Array [
                  Object {
                    "type": "text",
                    "value": "3",
                  },
                ],
                "isHeader": undefined,
                "type": "tableCell",
              },
            ],
            "type": "tableRow",
          },
        ],
        "type": "table",
      },
      Object {
        "children": Array [
          Object {
            "type": "text",
            "value": "Searching in ",
          },
          Object {
            "children": Array [
              Object {
                "type": "text",
                "value": "github",
              },
            ],
            "title": undefined,
            "type": "link",
            "url": undefined,
          },
          Object {
            "type": "text",
            "value": " or ",
          },
          Object {
            "children": Array [
              Object {
                "type": "text",
                "value": "github",
              },
            ],
            "title": undefined,
            "type": "link",
            "url": "https://github.com",
          },
          Object {
            "type": "text",
            "value": " for details.",
          },
        ],
        "type": "paragraph",
      },
      Object {
        "type": "thematicBreak",
      },
      Object {
        "children": Array [
          Object {
            "alt": "image",
            "src": "./fake.png",
            "title": undefined,
            "type": "image",
          },
          Object {
            "type": "text",
            "value": "
",
          },
          Object {
            "alt": "image",
            "src": undefined,
            "title": undefined,
            "type": "image",
          },
        ],
        "type": "paragraph",
      },
      Object {
        "children": Array [
          Object {
            "type": "text",
            "value": "Math",
          },
        ],
        "identifier": "heading-math",
        "level": 2,
        "type": "heading",
      },
      Object {
        "args": Object {},
        "lang": "math",
        "meta": "embed",
        "type": "codeEmbed",
        "value": "x^2 + y^2 = z^2
",
      },
      Object {
        "destination": "https://github.com",
        "identifier": "github",
        "label": "github",
        "title": undefined,
        "type": "definition",
      },
      Object {
        "destination": "./fake.png",
        "identifier": "fake-image",
        "label": "fake-image",
        "title": undefined,
        "type": "definition",
      },
    ],
    "type": "root",
  },
  "meta": Object {
    "definition": Object {
      "fake-image": Object {
        "destination": "./fake.png",
        "identifier": "fake-image",
        "label": "fake-image",
        "title": undefined,
      },
      "github": Object {
        "destination": "https://github.com",
        "identifier": "github",
        "label": "github",
        "title": undefined,
      },
    },
  },
  "toc": Object {
    "anchors": Array [
      Object {
        "children": Array [
          Object {
            "id": "heading-emphasis",
            "title": Array [
              Object {
                "type": "text",
                "value": "Emphasis",
              },
            ],
          },
          Object {
            "id": "heading-strong",
            "title": Array [
              Object {
                "type": "text",
                "value": "Strong",
              },
            ],
          },
          Object {
            "id": "heading-delete",
            "title": Array [
              Object {
                "type": "text",
                "value": "Delete",
              },
            ],
          },
          Object {
            "id": "heading-code",
            "title": Array [
              Object {
                "type": "text",
                "value": "Code",
              },
            ],
          },
          Object {
            "children": Array [
              Object {
                "id": "heading-xxx",
                "title": Array [
                  Object {
                    "type": "text",
                    "value": "xxx",
                  },
                ],
              },
              Object {
                "id": "heading-complex-header-~demo~",
                "title": Array [
                  Object {
                    "type": "inlineCode",
                    "value": "complex",
                  },
                  Object {
                    "type": "text",
                    "value": " ",
                  },
                  Object {
                    "children": Array [
                      Object {
                        "type": "text",
                        "value": "header",
                      },
                    ],
                    "type": "strong",
                  },
                  Object {
                    "type": "text",
                    "value": " ~demo~",
                  },
                ],
              },
            ],
            "id": "heading-list",
            "title": Array [
              Object {
                "type": "text",
                "value": "List",
              },
            ],
          },
        ],
        "id": "heading-paragraph",
        "title": Array [
          Object {
            "type": "text",
            "value": "Paragraph",
          },
        ],
      },
      Object {
        "children": Array [
          Object {
            "id": "heading-math",
            "title": Array [
              Object {
                "type": "text",
                "value": "Math",
              },
            ],
          },
        ],
        "id": "heading-table",
        "title": Array [
          Object {
            "type": "text",
            "value": "Table",
          },
        ],
      },
    ],
  },
}
`;
